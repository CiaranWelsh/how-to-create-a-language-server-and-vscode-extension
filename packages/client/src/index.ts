import {
  LanguageClient,
  LanguageClientOptions,
  ServerOptions
} from 'vscode-languageclient';

let client: LanguageClient;

export function activate() {
  const serverModule = require.resolve(
    '@donaldpipowitch/vscode-extension-server'
  );

  // Debug options for server are used when we launch the extension in debug mode
  const serverOptions: ServerOptions = {
    run: { module: serverModule },
    debug: {
      module: serverModule,
      options: { execArgv: ['--inspect=6009'] }
    }
  };

  const clientOptions: LanguageClientOptions = {
    documentSelector: [
      {
        pattern: '**/.vscode/extensions.json',
        scheme: 'file'
      }
    ]
  };

  client = new LanguageClient(
    'vscode-extensions-files',
    'VS Code Extension Client',
    serverOptions,
    clientOptions
  );

  // Starting the client will also launch the server.
  client.start();
}

export function deactivate() {
  if (client) {
    return client.stop();
  }
}
