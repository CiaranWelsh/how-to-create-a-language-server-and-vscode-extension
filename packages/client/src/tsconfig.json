{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "declarationDir": "../dist",
    "outDir": "../dist",
    "types": ["node"]
  },
  "references": [{ "path": "../../server/src" }]
}
