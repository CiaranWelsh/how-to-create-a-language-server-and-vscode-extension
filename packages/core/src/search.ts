import axios, { Canceler } from 'axios';

export { Canceler };

/**
 * In the `.vscode/extensions.json` we'll need to use `${publisher.publisherName}.${extensionName}`.
 *
 * @example
 * {
 *   "publisher": {
 *     "publisherId": "d16f4e39-2ffb-44e3-9c0d-79d873570e3a",
 *     "publisherName": "esbenp",
 *     "displayName": "Esben Petersen",
 *     "flags": "none"
 *   },
 *   "extensionId": "96fa4707-6983-4489-b7c5-d5ffdfdcce90",
 *   "extensionName": "prettier-vscode",
 *   "displayName": "Prettier - Code formatter",
 *   "flags": "validated, public",
 *   "lastUpdated": "2018-08-09T12:05:04.413Z",
 *   "publishedDate": "2017-01-10T19:52:02.703Z",
 *   "releaseDate": "2017-01-10T19:52:02.703Z",
 *   "shortDescription": "VS Code plugin for prettier/prettier",
 *   "deploymentType": 0
 * }
 */
export type Extension = {
  publisher: {
    publisherId: string;
    publisherName: string;
    displayName: string;
    flags: string;
  };
  extensionId: string;
  extensionName: string;
  displayName: string;
  flags: string;
  lastUpdated: string;
  publishedDate: string;
  releaseDate: string;
  shortDescription: string;
  deploymentType: number;
};

export type SearchRequest = {
  cancel: Canceler;
  request: Promise<Extension[] | void>;
};

export function search(value: string): SearchRequest {
  const { token, cancel } = axios.CancelToken.source();
  const options = {
    cancelToken: token,
    url:
      'https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery',
    method: 'post',
    headers: {
      accept: 'application/json;api-version=5.0-preview.1;excludeUrls=true'
    },
    data: {
      filters: [
        {
          criteria: [
            // which visual studio app? code
            { filterType: 8, value: 'Microsoft.VisualStudio.Code' },
            // our search value
            { filterType: 10, value }
          ],
          pageSize: 10,
          pageNumber: 1
        }
      ]
    }
  };
  const request = axios(options)
    .then(({ data }) => data.results[0].extensions as Extension[])
    .catch((error) => {
      if (!axios.isCancel(error)) {
        throw error;
      }
    });
  return { request, cancel };
}
