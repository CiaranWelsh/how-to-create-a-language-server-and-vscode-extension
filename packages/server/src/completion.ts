import { search, Canceler } from '@donaldpipowitch/vscode-extension-core';
import {
  TextDocuments,
  CompletionItemKind,
  Connection
} from 'vscode-languageserver';
import { parseTree, findNodeAtOffset, Node } from 'jsonc-parser';

let lastCancel: Canceler | null = null;

export function configureCompletion(
  connection: Connection,
  documents: TextDocuments
) {
  // The onCompletion handler provides the initial list of the completion items.
  // This is an example of the params passed to the handler:
  // {
  //   "textDocument": {
  //     "uri": "file:///Users/pipo/workspace/test-txt/test.txt"
  //   },
  //   "position": {
  //     "line": 0,
  //     "character": 10
  //   },
  //   "context": {
  //     "triggerKind": 1
  //   }
  // }
  connection.onCompletion(async ({ position, textDocument }) => {
    if (lastCancel) {
      lastCancel();
      lastCancel = null;
    }

    if (!textDocument.uri.endsWith('.vscode/extensions.json')) return;

    const document = documents.get(textDocument.uri);
    if (!document) return;

    const tree = parseTree(document.getText());
    const node = findNodeAtOffset(tree, document.offsetAt(position));

    if (!isExtensionValue(node)) return;

    // only search for queries with at least three characters
    if (node.value && node.value.length <= 2) return;

    const query = node.value;
    // connection.console.log(`You searched for ${query}.`);

    const { cancel, request } = search(query);
    lastCancel = cancel;

    const extensions = await request;
    lastCancel = null;

    if (!extensions) return;

    const completionItems = extensions.map(
      ({
        publisher: { publisherName },
        extensionName,
        displayName,
        shortDescription
      }) => ({
        // `"` hack, because JSON treats `"` as part of the completion here
        // see https://github.com/Microsoft/vscode-extension-samples/issues/93#issuecomment-429849514
        label: `"${publisherName}.${extensionName}`,
        kind: CompletionItemKind.Text,
        detail: displayName,
        documentation: shortDescription
      })
    );

    return completionItems;
  });
}

function isExtensionValue(node: Node | undefined): node is Node {
  // no node
  if (!node) return false;
  // not a string node
  if (node.type !== 'string') return false;
  // not within an array
  if (!node.parent) return false;
  if (node.parent.type !== 'array') return false;
  // not on a "recommendations" or "unwantedRecommendations" property
  if (!node.parent.parent) return false;
  if (node.parent.parent.type !== 'property') return false;
  if (
    node.parent.parent.children![0].value !== 'recommendations' &&
    node.parent.parent.children![0].value !== 'unwantedRecommendations'
  )
    return false;
  // not on an object
  if (!node.parent.parent.parent) return false;
  if (node.parent.parent.parent.type !== 'object') return false;
  // not the root
  if (node.parent.parent.parent.parent) return false;
  // whoohoo
  return true;
}
