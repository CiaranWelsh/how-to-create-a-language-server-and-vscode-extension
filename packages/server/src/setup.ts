import {
  createConnection,
  TextDocuments,
  ProposedFeatures
} from 'vscode-languageserver';

const connection = createConnection(ProposedFeatures.all);

// text document manager (supports full document sync only)
const documents = new TextDocuments();

connection.onInitialize(() => ({
  // tells the client what we support:
  // - full text sync
  // - code completion
  capabilities: {
    textDocumentSync: documents.syncKind,
    completionProvider: {
      resolveProvider: true
    }
  }
}));

// Listen for events
documents.listen(connection);
connection.listen();

export { connection, documents };
