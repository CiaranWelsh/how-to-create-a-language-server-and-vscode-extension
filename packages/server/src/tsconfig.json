{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "declarationDir": "../dist",
    "outDir": "../dist",
    "types": ["vscode", "node"]
  },
  "references": [{ "path": "../../core/src" }]
}
