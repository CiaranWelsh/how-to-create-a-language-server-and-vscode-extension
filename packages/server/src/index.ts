import { connection, documents } from './setup';
import { configureCompletion } from './completion';

configureCompletion(connection, documents);
