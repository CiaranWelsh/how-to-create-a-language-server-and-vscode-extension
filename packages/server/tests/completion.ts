import * as core from '@donaldpipowitch/vscode-extension-core';
import { configureCompletion } from '../src/completion';
import { prettier } from './__fixtures__/search';

test('should provide completion items', async () => {
  const search = jest.spyOn(core, 'search');
  let resolve: Function;
  const request = new Promise((_resolve) => (resolve = _resolve));
  const cancel = () => resolve();
  search.mockReturnValue({ cancel, request });

  const mockedConnection = { onCompletion: jest.fn() };
  const mockedDocuments = {
    get() {
      return {
        getText() {
          return `{
  "recommendations": ["prettier"]
}`;
        },
        offsetAt() {
          return 24; // the `"`, before `prettier"]`
        }
      };
    }
  };
  configureCompletion(mockedConnection as any, mockedDocuments as any);

  expect(mockedConnection.onCompletion).toHaveBeenCalledTimes(1);
  const completionHandler = mockedConnection.onCompletion.mock.calls[0][0];
  const itemsPromise = completionHandler({
    textDocument: { uri: '.vscode/extensions.json' }
  });
  resolve!(prettier);
  expect(await itemsPromise).toMatchSnapshot();
});
