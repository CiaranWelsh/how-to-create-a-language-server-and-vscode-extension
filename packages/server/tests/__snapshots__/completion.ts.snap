// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should provide completion items 1`] = `
Array [
  Object {
    "detail": "Prettier - Code formatter",
    "documentation": "VS Code plugin for prettier/prettier",
    "kind": 1,
    "label": "\\"esbenp.prettier-vscode",
  },
  Object {
    "detail": "Prettier Now",
    "documentation": "VS Code plugin for Prettier Miscellaneous / Code Formatter",
    "kind": 1,
    "label": "\\"remimarsal.prettier-now",
  },
  Object {
    "detail": "Prettier - JavaScript formatter",
    "documentation": "VS Code plugin for jlongster/prettier with tabs support",
    "kind": 1,
    "label": "\\"passionkind.prettier-vscode-with-tabs",
  },
  Object {
    "detail": "Prettier - JavaScript formatter",
    "documentation": "Fork of prettier-vscode: VS Code plugin for Skywalker13/prettier-space-parenthesis",
    "kind": 1,
    "label": "\\"schroeter.prettier-vscode-space-parenthesis",
  },
  Object {
    "detail": "Prettier - JavaScript formatter",
    "documentation": "prettier or prettier + standard --fix",
    "kind": 1,
    "label": "\\"bysabi.prettier-vscode-standard",
  },
  Object {
    "detail": "Prettier - JavaScript formatter",
    "documentation": "prettier or prettier + semistandard --fix",
    "kind": 1,
    "label": "\\"bysabi.prettier-vscode-semistandard",
  },
  Object {
    "detail": "Prettier Fuge Standard",
    "documentation": "prettier with fuge-standard",
    "kind": 1,
    "label": "\\"clmystes.prettier-fuge-standard",
  },
  Object {
    "detail": "Prettier-Standard - JavaScript formatter",
    "documentation": "VS Code plugin for prettier + standard",
    "kind": 1,
    "label": "\\"numso.prettier-standard-vscode",
  },
  Object {
    "detail": "Prettier-Standard - JavaScript formatter",
    "documentation": "VSCode plugin for prettier standard",
    "kind": 1,
    "label": "\\"iSayme.vscode-prettier-standard",
  },
  Object {
    "detail": "MaxLint",
    "documentation": "Maximum automatic linting by running lebab, Prettier, and ESLint",
    "kind": 1,
    "label": "\\"yaksoy.MaxLint",
  },
]
`;
